### ISSUE TITLE
className implementation to solve #27, #4, and #197

### ISSUE THREAD

**Issue by epiqueras** [zeit/styled-jsx#246]
Thanks for the hard work on this lib! It's a massive step forward from all the other inline styles approaches.

There are a couple of things that make me hesitant to adopt it in a big project, and I wanted to discuss the feasibility of a different approach to scoping styles.
1. Complication of targeting children components. #197 . I like to reuse components like custom dropdowns throughout an app and change their styles through an `extraClassName` prop.
2. Potential performance hits of using data attributes. #4 .

I think there is a relatively easy way to solve these two issues.

Instead of putting the hash as a data attribute we could concatenate it to the className prop of every element in the scope. This would also remove the need for a function like `_styledJsxPrefix` from #27 . We could also concatenate it to any string that matches a class selector in the styles. That way custom props like `extraClassNames` will still work. If `extraClassNames` is an expression then we can concatenate the styleId explicitly.

E.g.
```jsx
export default props => (
  <div className={props.someBool ? 'a' : 'b'}>
    <p>only this paragraph will get the style :)</p>
    <CustomComponent extraClassName="custom-class"/>
    <CustomComponent extraClassName={props.someBool ? 'custom-class' : 'c'}/>
    <CustomComponent extraClassName={props.someClass /*!styleId!*/}/>
    <style jsx>{`
      p {
        color: red;
      }

      .a {
        background: blue;
      }

      .b {
        background: green;
      }

      .custom-class {
        font-size: 14px;
      }
    `}</style>
  </div>
)
```
Becomes:
```jsx
export default props => (
  <div className={(props.someBool ? 'a' : 'b') + 'cn2o3j'}>
    <p className="cn2o3j">only this paragraph will get the style :)</p>
    <CustomComponent extraClassName="custom-class cn2o3j"/>
    <CustomComponent extraClassName={(props.someBool ? 'custom-class' : 'c') + 'cn2o3j'}/>
    <CustomComponent extraClassName={props.someClass + 'cn2o3j'}/>
    <_JSXStyle
      styleId='cn2o3j'
      css={`
        p.cn2o3j {
          color: red;
        }

        .a.cn2o3j {
          background: blue;
        }

        .b.cn2o3j {
          background: green;
        }

        .custom-class.cn2o3j {
          font-size: 14px;
        }
      `}
      />
  </div>
)
```

This approach would let you create classes that are scoped to the parent component but still affect children and it avoids the usage of data attributes.

Would this work?

**Comment by giuseppeg** [comment]
@epiqueras hi, thanks for your input! 

Switching from data attributes to classes is definitely something that I want to do. I just created an issue #247 for it. You can see why we would still need a function like `_styledJsxPrefix`.

Regarding the `extraClassNames` idea I think that #209 could somehow fix that issue, but yea potentially we could automatically pass the class to the components as `styled-jsx-parent` attribute (prop). And from a child component you could then grab it and set it as `className`.

One "problem" is that on every child component you would need to grab the class and set it (same with `extraClassName`). We you are trying to style 3rd parties components this is not going to work. That's why right now we recommend to use the `:global` selector https://github.com/zeit/styled-jsx#global-selectors.

**Comment by epiqueras** [comment]
@giuseppeg 

I looked at #247 . Couldn't we just append the hash during compile time to avoid so many function calls. The only place where `JSXStyle.addClass` would be needed would be where spread operators are used, and it would need a check to see if a className is present to avoid overwriting previous rest operators.
E.g.
```js
// To avoid (`{...props} {...rest}`) rest overwriting props.className
if (typeof target !== 'string') {
  if (target.className) ....
```
But, this would make it really difficult to append to props that aren't called className, but contain selector strings found inside the `style` tag, like in my first example. `JSXStyle.addClass` would have to compare every prop against every selector in the current component.
Is this even desired? I think in cases like this it would be better to leave it up to the user to explicitly append the hash. But, instead of doing it in runtime or using a function like #209 to create a uniqueClass, it could be done with a comment that gets compiled. E.g. `const className = 'props.className' /* styled-jsx */`. There could also be `/* not-styled-jsx */` to avoid the className concatenation.

This would also solve the issue of styling third party components.

So, maybe the best solution would be to only concatenate to className props and leave spread operators and custom prop names for the user to concatenate explicitly?

**Comment by giuseppeg** [comment]
> Couldn't we just append the hash during compile time to avoid so many function calls. 

Yes we can do that for the explicit `className` attribute (prop) eg. `<div className='foo' />`.

Regarding the spread conflict you are right, we should only set it when `target` has `className` (also always set `className` as first property in case none of the spread objects has it.

> But, this would make it really difficult to append to props that aren't called className, but contain selector strings found inside the style tag, like in my first example.

I don't think that we should do that, if anything we can pass an attribute `styled-jsx-parentClass` with the parent scoped class.

> Is this even desired? I think in cases like this it would be better to leave it up to the user to explicitly append the hash. But, instead of doing it in runtime or using a function like #209 to create a uniqueClass, it could be done with a comment that gets compiled. E.g. const className = 'props.className' /* styled-jsx */. There could also be /* not-styled-jsx */ to avoid the className concatenation.

this part is not clear to me, can you rephrase?

**Comment by giuseppeg** [comment]
@epiqueras updated the issue, let me know if I missed something.

**Comment by giuseppeg** [comment]
> What I meant is that maybe we should avoid looking in spread objects for className props and class selectors in custom prop names.

The reason for wrapping spreads with a fn is that one might do the following:

```jsx
({ unusedProp, ...rest }) => (
  <div {...rest} />
)
```

**Comment by epiqueras** [comment]
@giuseppeg 

There are probably three cases where that happens:
1. Custom component made by the user and styles for rest.className are in the parent.
  Here, you expect rest.className to not be concatenated. It wouldn't do much if it is concatenated because the selector that styles it is in the parent's style tag and uses the parent's styleId.

2. Custom component made by the user and styles for rest.className are in the custom component.
  Here the user would have to explicitly concatenate it using the comment format. Still better than a function call, perf-wise.

3. Third party component and styles are in the parent.
  Same as number 1. The className is already scoped to the the parent where it was defined. If you want to include it in different scopes you should do it explicitly through comment concatenation.

I'm just trying to figure out a way to keep it compile-time. We could also use some sort of global variable instead of the comment.

**Comment by giuseppeg** [comment]
if you don't use spreads a lot it is not going to be a problem for you :) But I will see if I can come up with a way to avoid the run time work

**Comment by epiqueras** [comment]
@giuseppeg 

What if instead of adding the function wrapper you add this:

Before:
```jsx
<Component {...props}/>
```

After:
```jsx
<Component {...props} className={props.className && (props.className + styleId)}/>
```

**Comment by giuseppeg** [comment]
Yep I thought of that too yesterday, moving `className` to the end and do an override sounds good. I'll update the issue.

**Comment by epiqueras** [comment]
👍

**Comment by giuseppeg** [comment]
@epiqueras thanks for you input :) I updated the issue, let me know if I missed something

**Comment by epiqueras** [comment]
@giuseppeg Looks great! I'll close this now.
